fgspectra/frequency.py

```diff
@@ -3,120 +3,122 @@
 Frequency-dependent foreground components.
 
 This module implements the frequency-dependent component of common foreground
-contaminants. This consist only of factorizable components -- i.e. the components
-in this module describe :math:`f(\nu)` for which the component X has spectrum
-
-.. math:: C_{\ell}^{X} = f(\nu) C^X_{0,\ell}
+contaminants. 
 
 This package draws inspiration from FGBuster (Davide Poletti and Josquin Errard)
 and BeFoRe (David Alonso and Ben Thorne).
 """
+
+from abc import ABC, abstractmethod
 import numpy as np
 from scipy import constants
 
 # TODO: we need to figure out the unit situation.
 
-class FrequencySpectrum:
+T_CMB = 2.725
+
+
+class FrequencySpectrum(ABC):
     """Base class for frequency dependent components."""
 
-    def __init__(self, sed_name=''):
-        """Initialize the component."""
+    @abstractmethod
+    def __call__(self, nu, *args):
+        # Return the evaluation of the SED
         pass
 
-    def __call__(self, nu, **kwargs):
-        """Calls the object's `sed(nu, **kwargs)` method."""
-        return self.sed(nu, **kwargs)
-
 
 class PowerLaw(FrequencySpectrum):
-    r"""
-    FrequencySpectrum for a power law.
-    .. math:: f(\nu) = (nu / nu_0)^{\beta}
+    r""" Power Law
 
-    Parameters
-    ----------
-    nu: float or array
-        Frequency in Hz.
-    beta: float
-        Spectral index.
-    nu0: float
-        Reference frequency in Hz.
-
-    Methods
-    -------
-    __call__(self, nu, beta)
-        return the frequency scaling given by a power law.
+    .. math:: f(\nu) = (\nu / nu_0)^{\beta}
     """
-    def sed(nu, beta, nu_0):
-        """Compute the SED with the given frequency and parameters."""
-        par = self.get_missing(kwargs)
-        return (nu / nu_0)**beta
+
+    def __call__(self, nu, beta, nu0):
+        """ Evaluation of the SED
+
+        Parameters
+        ----------
+        nu: float or array
+            Frequency in Hz.
+        beta: float or array
+            Spectral index.
+        nu0: float
+            Reference frequency in Hz.
+
+        Returns
+        -------
+        sed: ndarray
+            The last dimension is the frequency dependence.
+            The leading dimensions are the hypothetic dimensions of `beta`
+        """
+        beta = np.array(beta)[..., np.newaxis]
```

Comment:
Why do we need dimensions of beta? (sorry I posed this multiple times)

Reply:
I put in the examples the cases of use I had in mind. Please let me know if you find confusing to compute many SEDs at once. I very much prefer it because otherwise you would resort to more lines of code that are less flexible and are very likely to contain for loops.

Reply:
Ok let us discuss on the call. Initially I would be fine with this having dimension zero.